cltk/phonology/orthophonology.py

```diff
@@ -0,0 +1,709 @@
+	's' : 's',
+	'z' : 'z',
+	'ʃ' : 'sh',
+	'ç' : 'ch',
+	'x' : 'ch', # tough one -- like in Scottish loch
+	'y' : 'y',
+	'h' : 'h', 
+	'l' : 'l',
+	'l̥' : 'l',
+	'j' : 'y',
+	'w' : 'w',
+	'r' : 'r',
+	'r̥' : 'r',
+	'i' : 'i',
+	'i:': 'ee',
+	'y:': 'y',
+	'u' : 'u',
+	'u:': 'oo',
+	'e' : 'e',
+	'e:': 'ee',
+	'ø' : 'e',
+	'ø:': 'ee',
+	'o' : 'o',
+	'o:': 'oo',
+	'æ' : 'a',
+	'æ:': 'aa',
+	'ɑ' : 'o',
+	'ɑ:': 'oo',
+	'æɑ': 'ao',
+	'æ:ɑ': 'ao',
+	'eo': 'eo',
+	'e:o': 'eeo',
+	'iu': 'iu',
+	'i:u': 'iiu'
+}
+
+mde_phonotactics = [
+	(r'(^|(?<= ))hw', 'wh'),
+	(r'oo(.)(^|(?= ))', 'o\\1e')
+]
+
+
+class Orthophonology:
+	'''
+	The ortho-phonology of a language is described by:
+	* The inventory of all the phonemes of the language.
+	* A mapping of orthographic symbols to phonemes.
+	* mappings of orthographic symbols pairs to:
+	    * diphthongs
+	    * phonemes (i.e. digraphs)
+	* phonological rules for the contexutal transformation of phonological representations.
+
+	The class is very clearly aimed at alphabetic orthographies.  
+	Its usefulness for e.g. pictographic orthographies is questionable.
+	'''
+	def __init__(self, sound_inventory, alphabet, diphthongs, digraphs, to_modern = (ipa_to_mde, mde_phonotactics)):
+		self.sound_inventory = sound_inventory
+		self.alphabet = alphabet
+		self.diphthongs = diphthongs
+		self.digraphs = digraphs
+		self.di = {**self.diphthongs, **self.digraphs}
+		self.rules = []
+		self.to_modern = to_modern
+
+	def add_rule(self, rule):
+		'''
+		Adds a rule to the orthophonology.
+		The *order* in which rules are added is critcial, since the first rule that matches fires.'''
+		self.rules.append(rule)
+
+
+	# these are not static because language-specific subclasses probably need access to the sound inventory
+	def is_syllable_initial(self, phonemes, pos):
+		if pos == len(phonemes) - 1:
+			return False
+		# start of word is always syllable-initial, otherwise use SSP
+		return pos == 0 or \
+		(phonemes[pos - 1].is_more_sonorous(phonemes[pos]) and not phonemes[pos].is_more_sonorous(phonemes[pos + 1]))
+
+	def is_syllable_final(self, phonemes, pos):
+		# end of word is always syllable-final, otherwise use SSP
+		return pos == len(phonemes) - 1  or self.is_syllable_initial(phonemes, pos + 1)
+
+	@staticmethod
+	def _tokenize(text):
+		text = text.lower()
+		text = re.sub(r"[.\";,:\[\]()!&?‘]", "", text)
+		return text.split(' ')
+
+	def _position_phonemes(self, phonemes):
+		'''
+		Mark syllable boundaries, and, in future, other positional/suprasegmental features?
+		'''
+		for i in range(len(phonemes)):
+			phonemes[i] = PositionedPhoneme(phonemes[i])
+			phonemes[i].syllable_initial = self.is_syllable_initial(phonemes, i)
+			phonemes[i].syllable_final = self.is_syllable_final(phonemes, i)
+
+		return phonemes
+
+	def _find_sound(self, phoneme) :
+		for sound in self.sound_inventory:
+			if sound.is_equal(phoneme):
+				return sound
+		raise PhonemeNotFound(phoneme)
+
+
+	def transcribe_word(self, word):
+		'''

```

Comment:
OK.  Note that the algorithm is essentially the same as @clemsciences with just the addition of the digraphs.  And his approach is a totally reasonable application of phonological theory.  There's still the issue of whether the rule system should keep making passes until there are no further changes, rather than greedy rule application.  A phonologist I asked opined that in most cases a greedy, progressive method is fine.  We'll see in practice.